tracing/logger.go

```diff
@@ -0,0 +1,250 @@
+// It has also some convenience methods to connect to a standard cothority
+//installation, including putting the nodeName in the logging information.
+type TraceLogger struct {
+	defaultFields    map[string]string
+	logInfo          log.LoggerInfo
+	done             chan bool
+	hcn              newTrace
+	NoSingleSpans    bool
+	PrintSingleSpans int
+	entryPoints      []string
+	doneMsgs         []string
+	traces           map[string]*traceWrapper
+	logMutex         sync.Mutex
+}
+
+type newTrace = func(context.Context, string) (context.Context, Trace)
+
+// Log implements the Logger interface.
+func (l *TraceLogger) Log(level int, msg string) {
+	msg = strings.TrimSpace(msg)
+	tr, sw := l.getTraceSpan(level, msg, 7)
+	if tr != nil {
+		for k, v := range l.defaultFields {
+			tr.hcTrace.AddField(k, v)
+		}
+		var m runtime.MemStats
+		runtime.ReadMemStats(&m)
+		sw.hcSpan.AddField("memstats", m)
+		var gc debug.GCStats
+		debug.ReadGCStats(&gc)
+		sw.hcSpan.AddField("gcstats", gc)
+		ts, err := cpu.Times(false)
+		if err == nil {
+			sw.hcSpan.AddField("cpustat", ts)
+		}
+		ld, err := load.Avg()
+		if err == nil {
+			sw.hcSpan.AddField("load", ld)
+		}
+		us, err := disk.Usage(".")
+		if err == nil {
+			sw.hcSpan.AddField("disk-usage", us)
+		}
+		ioc, err := disk.IOCounters(".")
+		if err == nil {
+			sw.hcSpan.AddField("disk-iostat", ioc)
+		}
+		netio, err := net.IOCounters(false)
+		if err == nil {
+			sw.hcSpan.AddField("network-stat", netio)
+		}
+	}
+}
+
+// Close implements the Logger interface.
+func (l *TraceLogger) Close() {
+	close(l.done)
+}
+
+// GetLoggerInfo implements the Logger interface.
+func (l *TraceLogger) GetLoggerInfo() *log.LoggerInfo {
+	return &l.logInfo
+}
+
+// AddOnetDefaults sets a number of entry-points that are useful when
+//running a project using onet.
+// These entry-points allow to track quite accurately what is happening,
+//both for the servie-calls over websockets, as well as the protocol-messages.
+func (l *TraceLogger) AddOnetDefaults(si *network.ServerIdentity) {
+	l.AddEntryPoints(
+		"go.dedis.ch/onet/v3.wsHandler.ServeHTTP",
+		"go.dedis.ch/onet/v3/network.(*BlockingDispatcher).Dispatch",
+		"go.dedis.ch/onet/v3/network.(*RoutineDispatcher).Dispatch",
+		"go.dedis.ch/cothority/v3/blscosi/protocol.(*SubBlsCosi).dispatchLeaf",
+		"go.dedis.ch/onet/v3.(*TreeNodeInstance).dispatchMsgToProtocol",
+		"go.dedis.ch/onet/v3.(*Overlay).TransmitMsg",
+		"go.dedis.ch/onet/v3.(*TreeNodeInstance).dispatchMsgReader")
+	l.AddDoneMsgs("ws close", "done measuring")
+
+	l.defaultFields["nodeName"] = si.String()
+	l.defaultFields["nodeDescription"] = si.Description
+}
+
+// AddEntryPoints takes all given entry points and adds them to the internal
+// list.
+// Empty entry points are discarded.
+func (l *TraceLogger) AddEntryPoints(eps ...string) {
+	for _, ep := range eps {
+		if len(ep) > 0 {
+			l.entryPoints = append(l.entryPoints, ep)
+		}
+	}
```

Comment:
Because the `entryPoints` and `doneMsgs` are somewhat crucial I prefer having them private and then checking for empty strings before adding them.
Also, not failing here allows for a simpler usage for environment-variables by simply passing the `strings.Split` result.

Reply:
In `AddEnvironment`, you check for the content of `TRACING_PRINT_SINGLE_SPAN` and `TRACING_CREATE_SINGLE_SPANS`, shouldn't be the check for empty values be put there instead?
Maybe, someone wants to trace every function, thus using `""` (or `"."` btw) as a valid entrypoint.

Reply:
The `SINGLE_SPANS` are more of a mapping of log-messages to the tracing framework, and not really useful. As such they are not an `entry point`, as they will not be able to receive other spans. So it's two different things.